enhancements/installer/aws-outpost-remote-workers.md

```diff
@@ -0,0 +1,209 @@
+- Admin can follow documentation to customize installer generated manifests to deploy
+  OCP on AWS Outpost
+- As post installation step admin can configure ALB on AWS Outpost as workload ingress
+- Admin can check list of limitations or not supported features
+
+Phase 1:
+- Admin can use OCP installer to deploy OCP on AWS Outpost
+- After installation only gp2 storageclass is available and ingress is configured
+
+Phase 2:
+- Admin can use ROSA cli or UI to deploy OCP/Hypershift on AWS Outpost
+
+### Non-Goals
+
+We do not intend to implement AWS Outpost specific features in OCP as well us we
+do not plan to handle long lasting networking issues between regular AWS region andan outpost.
+
+## Proposal
+
+Phase 0:
+We want to provide documented manual procedure on how to deploy an OCP cluster in a regular
+AWS region with worker nodes running in AWS Outpost. The user needs to create VPC and
+subnets before the installation and provide them in the install-config to prevent VPC
+creation by IPI installer. The subnets need to be created both in the regular region and
+in the outpost (tagged as "kubernetes.io/cluster/<non-cluster-name>": "owned"). In the
+install-config only AWS region subnet needs to be used so NLB is created for the
+apiserver to use. During manifest stage it is needed to change workers' machineSet
+to use subnets created in the outpost. As the last modification the user needs to
+update MTU for the network provider due to AWS Outpost supporting only 1300 bytes. Once the
+cluster is deployed the user needs to install AWS load balancer operator and configure
+ALB ingress.
+
+Phase 1:
+Separate enhacement needed
+
+Phase 2:
+Separate enhacement needed
+
+### Workflow Description
+
+Phase 0:
+In order to start the cluster installation process the user needs to follow steps to
+configure AWS account and AWS Outpost instance types. Next the user needs to create
+VPC with subnets for both the AWS region and the outpost. The user needs to create
+cloud formation template with AWS outpost details and use aws client to run it.
+Next the user needs to create an install-config in the specified directory and modify it
+with AWS details according to documentation. Once it is ready the user needs to
+create manifests. MachineSet needs to be updated with AWS outpost subnet information
+and MTU. Now we can create the cluster and after the installation we need to install
+AWS load balancer operator and configure ALB based ingress.
+
+Phase 1:
+Separate enhacement needed
+
+Phase 2:
+Separate enhacement needed
+
+#### Variation [optional]
+
+N/A
+
+### API Extensions
+
+N/A - no api extensions are being introduced in this proposal
+
+### Implementation Details/Notes/Constraints [optional]
+
+NLB is not supported by AWS outpost and it is needed for apiserver to work correctly.
+That is why in Phase 0 we have decided to run master nodes in the regular AWS region and only
+workers in AWS outpost where we can configure ALB for workload ingress.
```

Comment:
For OCP, master is just a regular node with a taint. Therefore any workload can actually run there, esp. when user makes [masters schedulable using OCP API](https://docs.openshift.com/container-platform/4.11/nodes/nodes/nodes-nodes-managing.html#nodes-nodes-working-master-schedulable_nodes-nodes-jobs).

In that case, we may have an issue with PersistentVolumes, because they may have wrong / misleading  `nodeAffinity`.

### Regular AWS cloud

A PV that is provisioned for a regular AWS cloud will have this nodeAffinity:

```
  nodeAffinity:
    required:
      nodeSelectorTerms:
      - matchExpressions:
        - key: topology.ebs.csi.aws.com/zone
          operator: In
          values:
          - us-east-1a
```

Node in a regular cluster has this labels:

```
topology.ebs.csi.aws.com/zone: us-east-1a
```

### Outpost

Node in Outpost have these labels:
```
topology.ebs.csi.aws.com/zone: us-east-1a
topology.ebs.csi.aws.com/region: us-east-1
topology.ebs.csi.aws.com/partition: aws
topology.ebs.csi.aws.com/account-id: "320667403945"
topology.ebs.csi.aws.com/outpost-id: op-0a759e7fbceffcb86
```

Notice that the cloud PV requires *any* node that has a label `topology.ebs.csi.aws.com/zone` == `us-east-1a` and that matches also outpost nodes. Therefore kubernetes scheduler may put a pod that needs a cloud PV to an outpost node. Can an outpost node attach a cloud volume? How fast and how expensive it is?

 It seems to me the CSI driver never supported nodes in mixed envs and masters should be in the outpost too (or in HyperShift).

We will either need a strong documentation forbidding scheduling anything that needs PVs to masters (incl. e.g. registry, when (mis?)configured to use PVCs) or we need to fix the CSI driver.

Reply:
Hi @jsafrane ,
Thank you for pointing this out
I opened a ticket to have more investigation on this issue: https://issues.redhat.com/browse/ECOPROJECT-1088
This is also relevant to a cluster contains worker nodes both in Outposts and in the region.
Currently this scenario is not supported for this release, as it is not tested .

Reply:
It's surprising to me that these outpost nodes are commingled with nodes in the AWS zone, is there no way to define the outpost as its own zone? Is there work in the CSI driver to track outposts? What's EKS doing here?